### PR TITLE
Add rule to convert *.mikro-e files to *.hex

### PR DESCRIPTION
**Pull request by Avinash95**
this connects #55

### CODE REVIEW

File: platform/mikro-e/apps/test-button/Makefile [hunk 1]
```diff
@@ -1,6 +1,9 @@
 CONTIKI_PROJECT = test-button
-all: $(CONTIKI_PROJECT)
-
+ifndef TARGET
+TARGET = mikro-e
+endif
 CONTIKI = ../../../../
-CFLAGS += -I $(CONTIKI)/platform/mikro-e/
+CFLAGS += -I $(CONTIKI)/platform/$(TARGET)/
+all: test-button
```
Comment (abhijit-mahajani-imgtec) [inline platform/mikro-e/apps/test-button/Makefile]: why did you remove the $(CONTIKI_PROJECT) from all?

File: platform/mikro-e/apps/test-button/Makefile [hunk 2]
```diff
@@ -1,6 +1,9 @@
 CONTIKI_PROJECT = test-button
-all: $(CONTIKI_PROJECT)
-
+ifndef TARGET
```
Comment (abhijit-mahajani-imgtec) [inline platform/mikro-e/apps/test-button/Makefile]: Either if all these sample/test apps are not dependent on platform then it should not be under mikro-e, but if its dependent, then why can't we have TARGET=mikro-e unconditionally?

**Review by abhijit-mahajani-imgtec — CHANGES_REQUESTED** [review]

File: platform/mikro-e/apps/test-button/Makefile (hunk 1 above)
Comment (Avinash95) [reply platform/mikro-e/apps/test-button/Makefile]: because it's not needed

**Comment by mtusnio** [comment]
I've just tried making the sample-app, removed main.hex, then re-ran "make". Until you run make clean the makefile doesn't treat main.hex as a file it needs to regenerate.

File: platform/mikro-e/Makefile.mikro-e [hunk 3]
```diff
@@ -50,3 +54,21 @@ CFLAGS += -DRF_CHANNEL=$(CHANNEL)
...
 include $(CONTIKI)/cpu/pic32/Makefile.pic32
+
+CUSTOM_RULE_LINK=1
+
+### CPU-dependent cleanup files
+CLEAN += *.hex *.$(TARGET)
+
+.PHONY: bin2hex
+
+all: $(CONTIKI_PROJECT).$(TARGET) bin2hex
```
Comment (mtusnio) [inline platform/mikro-e/Makefile.mikro-e]: When using all you can do it much easier, you can get rid of the CUSTOM_RULE_LINK and the custom rule below and just do it like this (see full patch below):

all: $(CONTIKI_PROJECT).$(TARGET)
    xc32-bin2hex $(CONTIKI_PROJECT).mikro-e

This is the kind of a fix I was thinking of, it's just in this case we have to get rid of the dependency on CONTIKI_PROJECT. 

```diff --git a/platform/mikro-e/Makefile.mikro-e b/platform/mikro-e/Makefile.mikro-e
index d2bea39..6fdb602 100755
--- a/platform/mikro-e/Makefile.mikro-e
+++ b/platform/mikro-e/Makefile.mikro-e
@@ -55,20 +55,15 @@ CFLAGS += -DIEEE802154_CONF_PANID=$(PAN_ID)
 
 include $(CONTIKI)/cpu/pic32/Makefile.pic32
 
-CUSTOM_RULE_LINK=1
+#CUSTOM_RULE_LINK=1
 
 ### CPU-dependent cleanup files
 CLEAN += *.hex *.$(TARGET)
 
-.PHONY: bin2hex
-
-all: $(CONTIKI_PROJECT).$(TARGET) bin2hex
-
-$(CONTIKI_PROJECT).$(TARGET): $(CONTIKI_PROJECT).co $(PROJECT_OBJECTFILES) $(PROJECT_LIBRARIES) contiki-$(TARGET).a
-       $(TRACE_LD)
-       $(Q)$(LD) $(LDFLAGS) $(TARGET_STARTFILES) ${filter-out %.a,$^} ${filter %.a,$^} $(TARGET_LIBFILES) -o $@
-
-bin2hex:
-       xc32-bin2hex $(CONTIKI_PROJECT).$(TARGET)
-
+#%.$(TARGET): %.co $(PROJECT_OBJECTFILES) $(PROJECT_LIBRARIES) contiki-$(TARGET).a
+#      $(TRACE_LD)
+#      $(Q)$(LD) $(LDFLAGS) $(TARGET_STARTFILES) ${filter-out %.a,$^} \
+#              ${filter %.a,$^} $(TARGET_LIBFILES) -o $@
 
+all: $(CONTIKI_PROJECT).$(TARGET)
+       xc32-bin2hex $(CONTIKI_PROJECT).mikro-e
```

File: platform/mikro-e/Makefile.mikro-e [hunk 4]
```diff
@@ -50,3 +54,21 @@ CFLAGS += -DRF_CHANNEL=$(CHANNEL)
...
+
+$(CONTIKI_PROJECT).$(TARGET): $(CONTIKI_PROJECT).co $(PROJECT_OBJECTFILES) $(PROJECT_LIBRARIES) contiki-$(TARGET).a
+	$(TRACE_LD)
+	$(Q)$(LD) $(LDFLAGS) $(TARGET_STARTFILES) ${filter-out %.a,$^} ${filter %.a,$^} $(TARGET_LIBFILES) -o $@
+
+bin2hex:
```
Comment (mtusnio) [inline platform/mikro-e/Makefile.mikro-e]: Fails when make -j9 is used, breaks multithreaded builds.

**Review by mtusnio — CHANGES_REQUESTED** [review]
Breaks CI since it introduces a dependency on CONTIKI_PROJECT.

**Review by mtusnio — DISMISSED** [review]
The second commit seems to add more than the commit message implies, can you split it into separate commit(s)?

**Review by abhijit-mahajani-imgtec — APPROVED** [review]

**Review by abhijit-mahajani-imgtec — CHANGES_REQUESTED** [review]
Need to update the branch

**Review by abhijit-mahajani-imgtec — APPROVED** [review]

**Review by mtusnio — COMMENTED** [review]
According to this - https://www.gnu.org/software/make/manual/html_node/Force-Targets.html - you can do something along those lines to always force regeneration of the hex file:

```diff
diff --git a/platform/mikro-e/Makefile.mikro-e b/platform/mikro-e/Makefile.mikro-e
index 1b4fa63..8f45087 100755
--- a/platform/mikro-e/Makefile.mikro-e
+++ b/platform/mikro-e/Makefile.mikro-e
@@ -70,8 +70,10 @@ CUSTOM_RULE_LINK = 1
        $(Q)$(LD) $(LDFLAGS) $(TARGET_STARTFILES) ${filter-out %.a,$^} \
            ${filter %.a,$^} $(TARGET_LIBFILES) -o $@
 
-%.hex: %.bin
+%.hex: %.bin FORCE
        xc32-bin2hex $<
 
+FORCE:
+
 %.$(TARGET): %.bin %.hex
        mv $< $@
```
The downside is that the final .c file will be always compiled when you run "make", since it will force a build of the hex file, but that solves the issue of the hex file not rebuilding if removed.

**Review by mtusnio — APPROVED** [review]

**Pull request merged**